apis/apiextensions/v1/composition_types.go

```diff
@@ -328,25 +446,79 @@ func (c *Patch) applyFromFieldPathPatch(from, to runtime.Object) error { // noli
 	if err != nil {
 		return err
 	}
-	out := in
-	for i, f := range c.Transforms {
-		if out, err = f.Transform(out); err != nil {
-			return errors.Wrapf(err, errFmtTransformAtIndex, i)
-		}
+
+	// Apply transform pipeline
+	out, err := c.applyTransforms(in)
+	if err != nil {
+		return err
 	}
 
-	if u, ok := to.(interface{ UnstructuredContent() map[string]interface{} }); ok {
-		return fieldpath.Pave(u.UnstructuredContent()).SetValue(*c.ToFieldPath, out)
+	return patchFieldValueToObject(*c.ToFieldPath, out, to)
+}
+
+// applyCombineFromVariablesPatch patches the "to" resource, taking a list of
+// input variables and combining them into a single output value.
+// The single output value may then be further transformed if they are defined
+// on the patch.
+func (c *Patch) applyCombineFromVariablesPatch(from, to runtime.Object) error { // nolint:gocyclo
+	// Combine patch requires configuration
+	if c.Combine == nil {
+		return errors.Errorf(errRequiredField, "Combine", c.Type)
+	}
+	// Destination field path is required since we can't default to multiple
+	// fields.
+	if c.ToFieldPath == nil {
+		return errors.Errorf(errRequiredField, "ToFieldPath", c.Type)
 	}
 
-	toMap, err := runtime.DefaultUnstructuredConverter.ToUnstructured(to)
+	vl := len(c.Combine.Variables)
+
+	if vl < 1 {
+		return errors.New(errCombineRequiresVariables)
+	}
+
+	fromMap, err := runtime.DefaultUnstructuredConverter.ToUnstructured(from)
+	if err != nil {
+		return err
+	}
+
+	in := make([]interface{}, vl)
+
+	// Get value of each variable
+	// NOTE: This currently assumes all variables define a 'fromFieldPath'
+	// value. If we add new variable types, this may not be the case and
+	// this code may be better served split out into a dedicated function.
+	for i, sp := range c.Combine.Variables {
+		iv, err := fieldpath.Pave(fromMap).GetValue(sp.FromFieldPath)
+
+		// If any source field is not found, we will not
+		// apply the patch. This is to avoid situations
+		// where a combine patch is expecting a fixed
+		// number of inputs (e.g. a string format
+		// expecting 3 fields '%s-%s-%s' but only
+		// receiving 2 values).
+		if IsOptionalFieldPathNotFound(err, c.Policy) {
+			return nil
+		}
+		if err != nil {
+			return err
+		}
```

Comment:
Tried this to see and it wouldn't compile, I think because `IsOptionalFieldPathNotFound` isn't the same shape as a `resource.ErrorIs` due to the additional input argument 🤔 (`cannot use IsOptionalFieldPathNotFound(err, c.Policy) (value of type bool) as resource.ErrorIs value in argument to resource.Ignore`)

Reply:
Ah, that makes sense. No big deal - I wasn't super sure it was an improvement anyhow.